dll/win32/shlwapi/appcompat.c

```diff
@@ -0,0 +1,443 @@
+    { "autorun.exe", "4.10.1998", SHACF_ANSI },
+    { "autorun.exe", "4.00.950", SHACF_ANSI },
+    { "powerpnt.exe", MAJOR_VER_ONLY "8", SHACF_WIN95SHLEXEC },
+    { "msmoney.exe", "7.05.1107", SHACF_WIN95SHLEXEC },
+    { "soffice.exe", MAJOR_VER_ONLY "5", SHACF_STAROFFICE5PRINTER },
+    { "wpwin9.exe", MAJOR_VER_ONLY "9", SHACF_CORELINTERNETENUM },
+    { "qpw.exe", MAJOR_VER_ONLY "9", SHACF_CORELINTERNETENUM },
+    { "prwin9.exe", MAJOR_VER_ONLY "9", SHACF_CORELINTERNETENUM },
+    { "dad9.exe", MAJOR_VER_ONLY "9", SHACF_CORELINTERNETENUM },
+};
+
+// Window class name and compatibility flags
+typedef struct tagWNDCOMPATINFO
+{
+    PCSTR pszLengthAndClassName;
+    DWORD dwCompatFlags;
+} WNDCOMPATINFO, *PWNDCOMPATINFO;
+static WNDCOMPATINFO g_wndCompatInfo[] =
+{
+    // The first byte indicates the string length
+    { "\x09" "bosa_sdm_", 0x1000100 },
+    { "\x18" "File Open Message Window", 0x1000100 },
+};
+
+// Internal structure for SHLWAPI_WndCompatEnumProc
+typedef struct tagAPPCOMPATENUM
+{
+    PWNDCOMPATINFO pItems;
+    SIZE_T nItems;
+    DWORD dwProcessId;
+    INT iFound;
+} APPCOMPATENUM, *PAPPCOMPATENUM;
+
+static BOOL CALLBACK
+SHLWAPI_WndCompatEnumProc(_In_ HWND hWnd, _In_ LPARAM lParam)
+{
+    PAPPCOMPATENUM pEnum = (PAPPCOMPATENUM)lParam;
+
+    CHAR szClass[256];
+    if (!pEnum->nItems || !GetClassNameA(hWnd, szClass, _countof(szClass)))
+        return TRUE; // Ignore and continue
+
+    // Search for the target window in pEnum
+    const INT cchClass = lstrlenA(szClass);
+    for (UINT iItem = 0; iItem < pEnum->nItems; ++iItem)
+    {
+        PCSTR pszLengthAndClassName = pEnum->pItems[iItem].pszLengthAndClassName;
+
+        INT cchLength = pszLengthAndClassName[0]; // The first byte represents the length
+        if (cchClass < cchLength)
+            cchLength = cchClass; // Ignore trailing characters
+
+        // Compare the string
+        if (StrCmpNA(szClass, &pszLengthAndClassName[1], cchLength) == 0) // Class name matched
+        {
+            // Get the process ID
+            DWORD dwProcessId;
+            GetWindowThreadProcessId(hWnd, &dwProcessId);
+            if (dwProcessId == pEnum->dwProcessId) // Same process
+            {
+                pEnum->iFound = iItem; // Found
+                return FALSE; // Quit enumeration
+            }
+        }
+    }
+
+    return TRUE; // Continue enumeration
+}
+
+// English (US) UTF-16
+#define PRODUCT_VER_ENGLISH_US_UTF16    "\\StringFileInfo\\040904E4\\ProductVersion"
+// German (Germany) UTF-16
+#define PRODUCT_VER_GERMAN_UTF16        "\\StringFileInfo\\040704E4\\ProductVersion"
+// English (US) Western European
+#define PRODUCT_VER_ENGLISH_US_WE       "\\StringFileInfo\\040904B0\\ProductVersion"
+// English (US) Neutral
+#define PRODUCT_VER_ENGLISH_US_NEUTRAL  "\\StringFileInfo\\04090000\\ProductVersion"
+// Swedish (Sweden) Western European
+#define PRODUCT_VER_SWEDISH_WE          "\\StringFileInfo\\041D04B0\\ProductVersion"
```

Comment:
Why do we need these 5 hardcoded values, and not others? If so, isn't it possible to generate them at runtime instead?